### PR TITLE
Recover buffer position after reset

### LINKED ISSUE #69: Recover data on fram buffer after reset
**Describe the Enhancement**
Currently the fram is "cleared" between reset of the stm32 since the read/write addresses are stored in memory. By storing the addresses on the nvm fram chip, we would be able to reload them on device startup. This would restore the current state of the buffer and the board could pick up where it left off.

**Benefits**
Prevent less data loss during intermittent power

**Possible Drawbacks**
Less data would be able to stored in the buffer
More power usage from more I2C writes
More wear on the fram chip

**Additional Context**
The read and write addresses, and the buffer length are stored as `uint16_t`. That is a total of 6 bytes. I think that the benefits outweigh the downsides.

### PR DESCRIPTION
**Pull request by mirellameelo**
**Name/Affiliation/Title**
Ella Ferraz/UCSC

**Purpose of the PR**
Refer to issue #69 

**Development Environment**
Ubuntu 20.04.6 LTS
PlatformIO Core, version 6.1.16
Hardware: 2.2.1

**Test Procedure**
To verify the functionality of data retrieval from FRAM after power outages, follow these steps:

1. Run `example_retrieve_data`

- Upload and run `example_retrieve_data` on the STM32 board. This will initialize the system and write sample data to the FRAM buffer.

2. Simulate Power Outage

- After the initial setup, manually remove power from the STM32 board to simulate an unexpected power outage.
- Reconnect power to restart the board.
- **Alternatively:**

    Press the Restart button on the STM32 board to simulate a power cycle without physically disconnecting power.

Expected Behavior on Reboot

- Upon reboot, the program should retrieve the saved `read_addr`, `write_addr`, and `buffer_len` states from FRAM.
- Pressing the **Restart** button should add additional test data to the FRAM buffer, confirming that data integrity and buffer states are preserved across power cycles.

3. Unit Test

- The `test_LoadSaveBufferState` function verifies that `read_addr` and `write_addr` in FRAM accurately reflect buffer states before and after simulated power outages.

- This unit test ensures that the buffer state is correctly saved and reloaded, maintaining data continuity across restarts.

**Additional Context**
The unit test for this feature validates the implemented methods at a software level when the "Restart" button is clicked. It ensures that the buffer state (read_addr and write_addr) accurately reflects data continuity across simulated restarts.

For a full testing procedure, including power cycle simulation, please refer to the Test Procedure section.

**Task List**

- [ ] Update `CHANGELOG.md`
- [ ] Static code analysis passes
- [ ] All environments can be built
- [ ] All tests pass
- [ ] Clear documentation for new code
- [ ] Linting passes
- [ ] (If applicable) Version bump python library

**Relevant Issues**
Closes #69

### CODE REVIEW

File: stm32/lib/fram/src/fifo.c [hunk 1]
```diff
@@ -119,6 +123,105 @@ FramStatus FramBufferClear(void) {
 
   // reset buffer len
   buffer_len = 0;
+  FramSaveBufferState(read_addr, write_addr, buffer_len);
+
+  return FRAM_OK;
+}
+
+void print(const char *format, ...) {
```
Comment (jmadden173) [inline stm32/lib/fram/src/fifo.c:131]: Remove this function, use `APP_PRINTF` or `APP_LOG` instead

File: stm32/lib/fram/src/fifo.c [hunk 2]
```diff
@@ -119,6 +123,105 @@ FramStatus FramBufferClear(void) {
...
+    } else {
+      print("Buffer contains data. Ready to resume operations.\n");
+    }
+  }
+  return FRAM_OK;
+}
+
+FramStatus FramSaveBufferState(uint16_t read_addr, uint16_t write_addr,
+                               uint16_t buffer_len) {
+  FramStatus status;
```
Comment (jmadden173) [inline stm32/lib/fram/src/fifo.c:164]: Initialize a value

File: stm32/lib/fram/src/fifo.c [hunk 3]
```diff
@@ -119,6 +123,105 @@ FramStatus FramBufferClear(void) {
...
+    return status;
+  }
+
+  // Print a single message summarizing the saved buffer state
+  print(
+      "Buffer State Saved Successfully. Read Address: 0x%04X (%d), Write "
+      "Address: 0x%04X (%d), Buffer Length: %d\n",
+      read_addr, read_addr, write_addr, write_addr, buffer_len);
+
+  return FRAM_OK;
```
Comment (jmadden173) [inline stm32/lib/fram/src/fifo.c:198]: Return status and initialize on creation

File: stm32/test/test_fifo/test_fifo.c
Comment (jmadden173) [inline stm32/test/test_fifo/test_fifo.c:1]: Don't see a test for saving the buffer state. The test doesn't call the load. When you add a test for it you can put it before the load buffer state and test them in sequence. IE the first test saves data, the second clears the buffer, restores it, and checks that the data is still available.

File: stm32/lib/fram/include/fifo.h
Comment (jmadden173) [inline stm32/lib/fram/include/fifo.h:1]: The addresses for read, write, and len are defined "privately" in the fifo library while the buffer save/load functions require the addresses to be passed as parameters. The functions can reference the globally defined variables.

File: stm32/lib/fram/src/fifo.c
Comment (jmadden173) [inline stm32/lib/fram/src/fifo.c:1]: As a general practice don't print anything within libraries as it clutters the log of the LoRa firmware. The FramStatus codes give us a way to print out debug messages from the main code.

**Review by jmadden173 — CHANGES_REQUESTED** [review]
Main comments are:
- removing print statements from the library (use `APP_LOG` or `APP_PRINTF`) in the future
- Convert load/save functions to use globally defined address variables
- Adding test for the save method

File: stm32/test/test_fifo/test_fifo.c
Comment (mirellameelo) [reply stm32/test/test_fifo/test_fifo.c:1]: FramPut calls  FramSaveBufferState(). I'll add a comment.

**Comment by jmadden173** [comment]
Try merging upstream changes from `release/2.2.0` and it should fix the build errors.

**Review by jmadden173 — APPROVED** [review]
lgtm

**Pull request merged**